instrumentation/jsf/jsf-testing-common/src/main/groovy/BaseJsfTest.groovy

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import static io.opentelemetry.api.trace.Span.Kind.INTERNAL
+import static io.opentelemetry.instrumentation.test.utils.TraceUtils.basicSpan
+
+import io.opentelemetry.instrumentation.test.AgentTestRunner
+import io.opentelemetry.instrumentation.test.asserts.TraceAssert
+import io.opentelemetry.instrumentation.test.base.HttpServerTestTrait
+import io.opentelemetry.sdk.trace.data.SpanData
+import okhttp3.FormBody
+import okhttp3.HttpUrl
+import okhttp3.Request
+import okhttp3.RequestBody
+import okhttp3.Response
+import org.eclipse.jetty.annotations.AnnotationConfiguration
+import org.eclipse.jetty.server.Server
+import org.eclipse.jetty.util.resource.Resource
+import org.eclipse.jetty.webapp.WebAppContext
+import org.jsoup.Jsoup
+
+abstract class BaseJsfTest extends AgentTestRunner implements HttpServerTestTrait<Server> {
+
+  @Override
+  Server startServer(int port) {
+    List<String> configurationClasses = new ArrayList<>()
+    Collections.addAll(configurationClasses, WebAppContext.getDefaultConfigurationClasses())
+    configurationClasses.add(AnnotationConfiguration.getName())
+
+    WebAppContext webAppContext = new WebAppContext()
+    webAppContext.setContextPath(getContextPath())
+    webAppContext.setConfigurationClasses(configurationClasses)
+    // set up test application
+    webAppContext.setBaseResource(Resource.newSystemResource("test-app"))
+    // add additional resources for test app
+    Resource extraResource = Resource.newSystemResource("test-app-extra")
+    if (extraResource != null) {
+      webAppContext.getMetaData().addWebInfJar(extraResource)
+    }
+    webAppContext.getMetaData().getWebInfClassesDirs().add(Resource.newClassPathResource("/"))
+
+    def jettyServer = new Server(port)
+    jettyServer.connectors.each {
+      it.setHost('localhost')
+    }
+
+    jettyServer.setHandler(webAppContext)
+    jettyServer.start()
+
+    return jettyServer
+  }
+
+  @Override
+  void stopServer(Server server) {
+    server.stop()
+    server.destroy()
+  }
+
+  @Override
+  String getContextPath() {
+    return "/jetty-context"
+  }
+
+  def "test hello"() {
+    setup:
+    def url = HttpUrl.get(address.resolve("hello.jsf")).newBuilder().build()
+    def request = request(url, "GET", null).build()
+    Response response = client.newCall(request).execute()
+
+    expect:
+    response.code() == 200
+    response.body().string().trim() == "Hello"
+
+    and:
+    assertTraces(1) {
+      trace(0, 1) {
+        basicSpan(it, 0, getContextPath() + "/hello.xhtml", null)
+      }
+    }
+  }
+
+  def "test hello with faces url"() {
+    setup:
+    def url = HttpUrl.get(address.resolve("faces/hello.xhtml")).newBuilder().build()
```

Comment:
Nit: could be merged into a parametrized test with `test hello` -- this is basically the same scenario except for different URLs.

Reply:
done